File: modules/daily-native/android/src/main/java/expo/modules/daily/Brain.kt
package expo.modules.daily

import android.content.Context

/**
 * 返事を作る「頭脳」の選択。
 *   auto  : まず Gemini。使えないとき（回数制限・混雑・通信なし）は端末内AIで答える
 *   cloud : Gemini だけ
 *   device: 端末内AIだけ（オフラインでも動く）
 */
object Brain {

  fun canAnswer(ctx: Context, mode: String, apiKey: String): Boolean = when (mode) {
    "device" -> LocalModel.isReady(ctx)
    "cloud" -> apiKey.isNotBlank()
    else -> apiKey.isNotBlank() || LocalModel.isReady(ctx)
  }

  fun missingMessage(mode: String): String =
    if (mode == "device") "端末内AIのモデルがありません。アプリの設定からダウンロードしてください。"
    else "アプリを開いて、設定でジェミニのAPIキーを入れてください。"

  fun ask(
    ctx: Context,
    mode: String,
    apiKey: String,
    model: String,
    systemPrompt: String,
    history: List<GeminiClient.Turn>
  ): String {
    val localReady = LocalModel.isReady(ctx)

    if (mode == "device") {
      if (!localReady) throw GeminiClient.GeminiException(missingMessage(mode))
      return LocalLlm.ask(ctx, systemPrompt, history)
    }
    if (mode == "cloud") {
      return GeminiClient.ask(apiKey, model, systemPrompt, history)
    }

    // auto
    if (apiKey.isBlank()) {
      if (!localReady) throw GeminiClient.GeminiException(missingMessage(mode))
      return LocalLlm.ask(ctx, systemPrompt, history)
    }
    return try {
      GeminiClient.ask(apiKey, model, systemPrompt, history)
    } catch (e: GeminiClient.GeminiException) {
      if (!localReady) throw e
      LocalLlm.ask(ctx, systemPrompt, history) // Gemini が使えないので、端末内AIで答える
    }
  }
}

File: src/brain.ts
import { askLocalModel, getLocalModelStatus } from '../modules/daily-native';
import { askGemini, type ChatTurn } from './gemini';
import type { Settings } from './settings';

type Brain = Settings['brain'];

const localReady = () => getLocalModelStatus().state === 'ready';

/** いまの設定で、返事を作れる状態か（APIキーの有無・端末内AIのダウンロード状況） */
export function canAnswer(brain: Brain, apiKey: string): boolean {
  if (brain === 'device') return localReady();
  if (brain === 'cloud') return apiKey.length > 0;
  return apiKey.length > 0 || localReady();
}

export function missingMessage(brain: Brain): string {
  return brain === 'device'
    ? '端末内AIのモデルがありません。右上の「設定」からダウンロードしてください。'
    : '右上の「設定」から Gemini API キーを入力してください。';
}

/**
 * 返事を作る。
 *   auto  : まず Gemini。使えないとき（回数制限・混雑・通信なし）は端末内AI
 *   cloud : Gemini だけ
 *   device: 端末内AIだけ
 */
export async function askBrain(params: {
  brain: Brain;
  apiKey: string;
  model: string;
  systemPrompt: string;
  history: ChatTurn[];
  signal?: AbortSignal;
}): Promise<string> {
  const { brain, apiKey, model, systemPrompt, history, signal } = params;
  const local = () => askLocalModel(systemPrompt, history);

  if (brain === 'device') return local();
  if (brain === 'cloud') return askGemini({ apiKey, model, systemPrompt, history, signal });

  if (!apiKey) return local();
  try {
    return await askGemini({ apiKey, model, systemPrompt, history, signal });
  } catch (e) {
    if (signal?.aborted || !localReady()) throw e;
    return local(); // Gemini が使えないので、端末内AIで答える
  }
}
